tools/querytee/response_comparator.go

```diff
@@ -0,0 +1,208 @@
+package querytee
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+// ResponsesComparatorFunc helps with comparing different responses from various routes.
+type ResponsesComparatorFunc func(expected, actual []byte) error
+
+// SamplesComparatorFunc helps with comparing different types of samples coming from /api/v1/query and /api/v1/query_range routes.
+type SamplesComparatorFunc func(expected, actual json.RawMessage) error
+
+type SamplesResponse struct {
+	Status string
+	Data   struct {
+		ResultType string
+		Result     json.RawMessage
+	}
+}
+
+var samplesComparator = map[string]SamplesComparatorFunc{
+	"matrix": compareMatrix,
+	"vector": compareVector,
+	"scalar": compareScalar,
+}
+
+// RegisterSamplesComparator helps with registering custom sample types
+func RegisterSamplesComparator(samplesType string, comparator SamplesComparatorFunc) {
+	samplesComparator[samplesType] = comparator
+}
+
+func CompareSamplesResponse(expectedResponse, actualResponse []byte) error {
+	var expected, actual SamplesResponse
+
+	err := json.Unmarshal(expectedResponse, &expected)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(actualResponse, &actual)
+	if err != nil {
+		return err
+	}
+
+	if expected.Status != actual.Status {
+		return fmt.Errorf("expected status %s but got %s", expected.Status, actual.Status)
+	}
+
+	if expected.Data.ResultType != actual.Data.ResultType {
+		return fmt.Errorf("expected resultType %s but got %s", expected.Data.ResultType, actual.Data.ResultType)
+	}
+
+	comparator, ok := samplesComparator[expected.Data.ResultType]
+	if !ok {
+		return fmt.Errorf("resultType %s not registered for comparison", expected.Data.ResultType)
+	}
+
+	return comparator(expected.Data.Result, actual.Data.Result)
+}
+
+type matrixResult struct {
+	Result []struct {
+		Metric map[string]string
+		Values [][]interface{}
+	}
+}
```

Comment:
Would it make sense to reuse existing data types from `prometheus/common/model` package here?